ir/function.cpp

```diff
@@ -289,57 +289,94 @@ void CFG::printDot(ostream &os) const {
   os << "}\n";
 }
 
-// Requires that CFG::edge_iterator visits each predecessor of a node before
-// the node itself and that the CFG is a DAG. Constructs dominators by: 
-// dom(x) = {intersect(dom(pred1), dom(pred2), ...)} U {x}. The result is a 
-// vector of dominator bb's of a given bb ordered by depth in CFG.
+// Cooper, Keith D.; Harvey, Timothy J.; and Kennedy, Ken (2001). 
+// A Simple, Fast Dominance Algorithm
+// http://www.cs.rice.edu/~keith/EMBED/dom.pdf
 void DomTree::buildDominators() {
-  for (auto [src, dst, instr] : cfg) {
+  unsigned bb_count = f.getBBs().size();
+
+  // build predecessor relationships
+  for (const auto &[p, b, instr] : cfg) {
     (void)instr;
 
-    // Make sure each bb dominates itself.
-    std::vector<const BasicBlock*> &srcDoms = (dominators.try_emplace(&src,
-                      std::vector<const BasicBlock*>{&src})).first->second;
-    std::vector<const BasicBlock*> &dstDoms = (dominators.try_emplace(&dst,
-                      std::vector<const BasicBlock*>{&dst})).first->second;
-
-    if (dstDoms.size() > 1) {
-      std::vector<const BasicBlock*> doms_copy = dominators[&dst];
-      dstDoms.clear();
-
-      // get the intersection of all predecessor dominator sets of dst
-      std::vector<const BasicBlock*>::const_iterator srcIt, dstIt;
-      srcIt = srcDoms.begin();
-      dstIt = doms_copy.begin();
-      while (srcIt != srcDoms.end() && dstIt != doms_copy.end()) {
-        if (*srcIt == *dstIt) {
-          dstDoms.push_back(*srcIt);
-          ++srcIt;
-          ++dstIt;
-        } else {
-          break;
+    auto &b_node = doms[&b];
+    if (!b_node.get())
+      b_node = make_unique<DomTreeNode>(b);
+    
+    auto &pred = doms[&p];
+    if (!pred.get())
+      pred = make_unique<DomTreeNode>(p);
+
+    b_node->preds.emplace_back(pred.get());
+  }
+
+  unsigned i = bb_count-1;
+  DomTreeNode *new_idom;
+  bool changed = true;
+  while (changed) {
+    changed = false;
+    for (auto &b : f.getBBs()) {
+      auto &b_node = doms[b];
+      b_node->order = i;
+      if (i == bb_count-1)
+        b_node->dominator = b_node.get();
+      --i;
```

Comment:
I have not been able to to get the index of a basic block in f.getBBs() from the loop above "for (const auto &[p, b, instr] : cfg) {" with pointer arithmetic, which would allow moving this initialization code to the first loop and improve efficiency for functions with loops.